File: ts/src/env.ts
import { EntryconfError } from "./errors.ts";
import { stripBom } from "./parse.ts";

const NAME_RE = /^[A-Za-z_][A-Za-z0-9_]*$/;

/** A process-environment lookup; `undefined` means unset. */
export type EnvLookup = (name: string) => string | undefined;

interface Definition {
  value: string;
  file: string;
}

/**
 * Parse one `*.env` file: a strict subset of dotenv (SPEC §4). Every line is
 * blank, a `#` comment, or `NAME=value`; anything else is E_PARSE.
 */
export function parseEnvFile(text: string, path: string): Map<string, string> {
  const defs = new Map<string, string>();
  const lines = stripBom(text).split("\n");
  for (let n = 0; n < lines.length; n++) {
    const raw = lines[n].replace(/\r$/, "");
    const line = raw.trim();
    if (line === "" || line.startsWith("#")) continue;

    const eq = line.indexOf("=");
    const name = eq === -1 ? "" : line.slice(0, eq);
    if (eq === -1 || !NAME_RE.test(name)) {
      throw new EntryconfError(
        "E_PARSE",
        `${path}:${n + 1}: expected a blank line, a "#" comment, or NAME=value`,
      );
    }
    if (defs.has(name)) {
      throw new EntryconfError(
        "E_ENV_CONFLICT",
        `${name} is defined more than once in ${path}`,
      );
    }
    defs.set(name, unquote(line.slice(eq + 1).trim()));
  }
  return defs;
}

function unquote(value: string): string {
  if (value.length >= 2) {
    const first = value[0];
    if ((first === '"' || first === "'") && value.endsWith(first)) {
      return value.slice(1, -1);
    }
  }
  return value;
}

/**
 * The single global variable namespace (SPEC §4): `*.env` definitions with the
 * process environment on top. Every lookup is remembered (SPEC §10.6
 * "variables"), and `where` reports which layer supplies a name.
 */
export class Vars {
  readonly files: Map<string, string>;
  /** name -> path of the `*.env` file defining it */
  readonly origin: Map<string, string>;
  readonly proc: EnvLookup;
  readonly used = new Set<string>();

  constructor(files: Map<string, string>, origin: Map<string, string>, proc: EnvLookup) {
    this.files = files;
    this.origin = origin;
    this.proc = proc;
  }

  get(name: string): string | undefined {
    this.used.add(name);
    const fromProcess = this.proc(name);
    if (fromProcess !== undefined) return fromProcess;
    return this.files.get(name);
  }

  /** "process", "file" (with the `*.env` path), or "" when unset. */
  where(name: string): { origin: "process" | "file" | ""; file: string } {
    if (this.proc(name) !== undefined) return { origin: "process", file: "" };
    const file = this.origin.get(name);
    if (file !== undefined) return { origin: "file", file };
    return { origin: "", file: "" };
  }
}

/**
 * Merge the `*.env` peers: all files are unordered peers, so a name defined
 * twice is a conflict. Returns the values and which file defines each.
 */
export function mergeEnvFiles(
  envFiles: { path: string; text: string }[],
): { files: Map<string, string>; origin: Map<string, string> } {
  const defs = new Map<string, Definition>();
  for (const file of envFiles) {
    for (const [name, value] of parseEnvFile(file.text, file.path)) {
      const existing = defs.get(name);
      if (existing) {
        throw new EntryconfError(
          "E_ENV_CONFLICT",
          `${name} is defined in both ${existing.file} and ${file.path}`,
        );
      }
      defs.set(name, { value, file: file.path });
    }
  }
  const files = new Map<string, string>();
  const origin = new Map<string, string>();
  for (const [name, def] of defs) {
    files.set(name, def.value);
    origin.set(name, def.file);
  }
  return { files, origin };
}

/** Kept for source compatibility with 0.2.0 internals: the merged namespace as a Map. */
export function buildNamespace(
  envFiles: { path: string; text: string }[],
  processEnv: Record<string, string | undefined>,
): Map<string, string> {
  const { files } = mergeEnvFiles(envFiles);
  for (const [name, value] of Object.entries(processEnv)) {
    if (value !== undefined) files.set(name, value);
  }
  return files;
}

File: go/env.go
package entryconf

import (
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// SPEC §4: NAME matches [A-Za-z_][A-Za-z0-9_]*.
var envNameRe = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

// envSource is the process-environment seam. The public Load uses the real
// process environment; the conformance harness injects a map instead so that
// no variable named by a fixture can leak in from the outside.
type envSource func(name string) (string, bool)

// vars is the single global variable namespace for a whole include tree
// (SPEC §4): the config directory's own *.env files, with the process
// environment taking precedence.
type vars struct {
	files  map[string]string
	origin map[string]string // name -> path of the *.env file defining it
	proc   envSource
	used   map[string]bool // every name looked up (SPEC §10.6 "variables")
}

func (v *vars) lookup(name string) (string, bool) {
	v.used[name] = true
	if val, ok := v.proc(name); ok {
		return val, true
	}
	val, ok := v.files[name]
	return val, ok
}

// where reports which layer of SPEC §4 supplies name: "process", "file"
// (with the *.env file's path), or "" when it is unset.
func (v *vars) where(name string) (origin, file string) {
	if _, ok := v.proc(name); ok {
		return "process", ""
	}
	if path, ok := v.origin[name]; ok {
		return "file", path
	}
	return "", ""
}

// loadEnvFiles reads every *.env file directly in dir (non-recursive). The
// files are unordered peers: a name defined twice, in one file or across two,
// is E_ENV_CONFLICT. The second map records which file defines each name.
func (l *loader) loadEnvFiles(dir string) (map[string]string, map[string]string, error) {
	names, err := l.src.envFileNames(dir)
	if err != nil {
		return nil, nil, wrapf(CodeParse, err, "cannot read config directory %q", dir)
	}
	sort.Strings(names) // deterministic error messages only; files are peers

	values := make(map[string]string)
	origin := make(map[string]string)
	for _, name := range names {
		path := filepath.Join(dir, name)
		data, err := l.src.readFile(path)
		if err != nil {
			return nil, nil, wrapf(CodeParse, err, "cannot read env file %q", path)
		}
		if err := checkUTF8(path, data); err != nil {
			return nil, nil, err
		}
		fileVals, err := parseEnvFile(path, string(data))
		if err != nil {
			return nil, nil, err
		}
		if l.rec != nil {
			l.rec.record(path, data, "env", nil)
		}
		for _, kv := range fileVals {
			if prev, dup := origin[kv.name]; dup {
				return nil, nil, errf(CodeEnvConflict,
					"variable %q defined in both %q and %q", kv.name, prev, path)
			}
			origin[kv.name] = path
			values[kv.name] = kv.value
		}
	}
	return values, origin, nil
}

type envPair struct{ name, value string }

// parseEnvFile implements the strict dotenv subset of SPEC §4. Anything that
// is not blank, a comment, or NAME=value is E_PARSE.
//
// The whole line is trimmed before classification, so indentation and
// trailing whitespace are fine; the name itself is then taken verbatim, so
// whitespace between NAME and "=" ("FOO = bar") is E_PARSE.
func parseEnvFile(path, text string) ([]envPair, error) {
	var out []envPair
	seen := make(map[string]int)
	for i, raw := range strings.Split(text, "\n") {
		lineNo := i + 1
		line := strings.TrimSpace(strings.TrimSuffix(raw, "\r"))
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		eq := strings.IndexByte(line, '=')
		if eq < 0 {
			return nil, errf(CodeParse, "%s:%d: not a NAME=value line: %q", path, lineNo, line)
		}
		name := line[:eq]
		if !envNameRe.MatchString(name) {
			return nil, errf(CodeParse, "%s:%d: invalid variable name %q", path, lineNo, name)
		}
		if prev, dup := seen[name]; dup {
			return nil, errf(CodeEnvConflict,
				"variable %q defined twice in %q (lines %d and %d)", name, path, prev, lineNo)
		}
		seen[name] = lineNo
		out = append(out, envPair{name: name, value: unquoteEnvValue(strings.TrimSpace(line[eq+1:]))})
	}
	return out, nil
}

// unquoteEnvValue strips one layer of matching single or double quotes. No
// escape processing: this is a strict subset of dotenv, all values are strings.
func unquoteEnvValue(v string) string {
	if len(v) >= 2 {
		q := v[0]
		if (q == '\'' || q == '"') && v[len(v)-1] == q {
			return v[1 : len(v)-1]
		}
	}
	return v
}
